homeassistant/components/sensor/metlink.py

```diff
@@ -0,0 +1,201 @@
+"""
+Support for Metlink's public transport.
+Bus, Train, Ferry. Wellington / Whanganui a Tara
+"""
+
+import logging
+from datetime import timedelta, datetime
+
+import requests
+
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ATTRIBUTION = "Data provided by metlink.org.nz"
+ICON = 'mdi:bus'
+STOP_URL = "https://www.metlink.org.nz/api/v1/StopDepartures/{stop_number}"
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Get the Metlink public transport sensor."""
+    # journal contains [0] Station ID start, [1] Station ID destination
+    # [2] Station name start, and [3] Station name destination
+    stop_number = config.get('stop_number')
+    route_number = config.get('route_number')
+    add_devices([MetlinkSensor(stop_number, route_number)])
+
+
+def pretty_timestamp(t):
+    #
+    # python2's strptime doesn't handle %z so here's a workaround, from:
+    #   http://stackoverflow.com/questions/1101508/how-to-parse-dates-with-0400-timezone-string-in-python/23122493#23122493
+    #
+    ret = datetime.strptime(t[0:16], '%Y-%m-%dT%H:%M')
+    if t[18] == '+':
+        ret += timedelta(hours=int(t[19:22]), minutes=int(t[23:]))
+    elif t[18] == '-':
+        ret -= timedelta(hours=int(t[19:22]), minutes=int(t[23:]))
+
+    return ret.strftime("%I:%M%p %a")
+
+
+class MetlinkSensor(Entity):
+    """Implementation of an Metlink public transport sensor."""
+
+    def __init__(self, stop_number, route_number):
+        """Initialize the sensor."""
+        self.stop_number = stop_number
+        self.route_number = route_number
+        self._name = 'stop_{stop_number}_route_{route_number}'.format(
+            stop_number=stop_number,
+            route_number=route_number
+        )
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return "Route {}".format(self.route_number)
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        # return self.metlink_stop.services[0].departure_status
+        service = self.metlink_stop.next_service(
+            route_number=self.route_number)
+        if service and service.is_real_time:
+            return int(service.departure_seconds / 60)
+        elif service and service.display_departure:
+            return pretty_timestamp(service.display_departure)
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attributes = {
+            'Stop': self.stop_number,
+            'Route': self.route_number,
+            'StopName': self.metlink_stop.stop_name,
+            'Latitude': self.metlink_stop.latitude,
```

Comment:
If you import `ATTR_LATITUDE` and `ATTR_LONGITUDE` from `homeassistant.const` and use those as the keys for latitude and longitude, the stop will also show up on the map in the frontend 👍